File: IAIso-v5.0/core/iaiso-csharp/src/Iaiso.Middleware/Cohere.cs
using System.Collections.Generic;
using System.Text.Json.Nodes;
using Iaiso.Core;

namespace Iaiso.Middleware.Cohere;

/// <summary>IAIso wrapper for Cohere chat.</summary>
public interface ICohereClient
{
    CohereResponse Chat(JsonObject @params);
}

public sealed class BilledUnits
{
    public long InputTokens { get; set; }
    public long OutputTokens { get; set; }
}

public sealed class CohereMeta
{
    public BilledUnits? Tokens { get; set; }
    public BilledUnits? BilledUnits { get; set; }
}

public sealed class CohereToolCall
{
    public string Name { get; }
    public CohereToolCall(string name) { Name = name; }
}

public sealed class CohereResponse
{
    public string Model { get; }
    public CohereMeta Meta { get; }
    public IReadOnlyList<CohereToolCall> ToolCalls { get; }
    public CohereResponse(string model, CohereMeta? meta, IReadOnlyList<CohereToolCall>? toolCalls)
    {
        Model = model;
        Meta = meta ?? new CohereMeta();
        ToolCalls = toolCalls ?? new List<CohereToolCall>();
    }
}

public sealed class CohereOptions
{
    public bool RaiseOnEscalation { get; set; }
    public static CohereOptions Defaults() => new();
}

public sealed class BoundedCohereClient
{
    private readonly ICohereClient _raw;
    private readonly BoundedExecution _execution;
    private readonly CohereOptions _opts;

    public BoundedCohereClient(ICohereClient raw, BoundedExecution execution, CohereOptions? opts = null)
    {
        _raw = raw;
        _execution = execution;
        _opts = opts ?? CohereOptions.Defaults();
    }

    public CohereResponse Chat(JsonObject @params)
    {
        var pre = _execution.Check();
        if (pre == StepOutcome.Locked) throw new MiddlewareException.Locked();
        if (pre == StepOutcome.Escalated && _opts.RaiseOnEscalation)
            throw new MiddlewareException.EscalationRaised();

        CohereResponse resp;
        try { resp = _raw.Chat(@params); }
        catch (System.Exception e) { throw new MiddlewareException.Provider(e.Message, e); }

        var b = resp.Meta.Tokens ?? resp.Meta.BilledUnits;
        long tokens = b is not null ? b.InputTokens + b.OutputTokens : 0;
        long toolCalls = resp.ToolCalls.Count;
        string model = string.IsNullOrEmpty(resp.Model) ? "unknown" : resp.Model;
        _execution.RecordStep(new StepInput
        {
            Tokens = tokens,
            ToolCalls = toolCalls,
            Tag = $"cohere.chat:{model}",
        });
        return resp;
    }
}

File: IAIso-v5.0/core/iaiso-ruby/lib/iaiso/middleware/cohere.rb
# frozen_string_literal: true

require_relative "errors"
require_relative "../core/pressure_engine"
require_relative "../core/step_outcome"

module IAIso
  module Middleware
    module Cohere
      BilledUnits = Data.define(:input_tokens, :output_tokens) do
        def self.empty = new(input_tokens: 0, output_tokens: 0)
      end
      Meta = Data.define(:tokens, :billed_units) do
        def self.empty = new(tokens: nil, billed_units: nil)
      end
      ToolCall = Data.define(:name)
      Response = Data.define(:model, :meta, :tool_calls)

      Options = Data.define(:raise_on_escalation) do
        def self.defaults = new(raise_on_escalation: false)
      end

      class BoundedClient
        def initialize(raw:, execution:, options: Options.defaults)
          @raw = raw
          @execution = execution
          @options = options
        end

        def chat(params)
          pre = @execution.check
          raise LockedError if pre == IAIso::Core::StepOutcome::LOCKED
          if pre == IAIso::Core::StepOutcome::ESCALATED && @options.raise_on_escalation
            raise EscalationRaisedError
          end
          resp =
            begin
              @raw.chat(params)
            rescue StandardError => e
              raise ProviderError, e.message
            end
          billed = resp.meta.tokens || resp.meta.billed_units
          tokens = billed.nil? ? 0 : (billed.input_tokens.to_i + billed.output_tokens.to_i)
          tool_calls = resp.tool_calls.size
          model = resp.model.to_s.empty? ? "unknown" : resp.model
          @execution.record_step(IAIso::Core::StepInput.build(
            tokens: tokens, tool_calls: tool_calls,
            tag: "cohere.chat:#{model}",
          ))
          resp
        end
      end
    end
  end
end
